[HLSL][SPIRV][DXIL] Implement `dot4add_i8packed` intrinsic

inbelic opened the pull request:
- create a clang built-in in Builtins.td
    - link dot4add_i8packed in hlsl_intrinsics.h
    - add lowering to spirv backend through expansion of operation as OPSDot is missing up to SPIRV 1.6 in SPIRVInstructionSelector.cpp
    - add lowering to spirv backend using OpSDot in applicable SPIRV version or if SPV_KHR_integer_dot_product is enabled
    - add dot4add_i8packed intrinsic to IntrinsicsDirectX.td and mapping to DXIL.td op Dot4AddI8Packed

    - add tests for HLSL intrinsic lowering to dx/spv intrinsic in dot4add_i8packed.hlsl
    - add tests for sema checks in dot4add_i8packed-errors.hlsl
    - add test of spir-v lowering in SPIRV/dot4add_i8packed.ll
    - add test to dxil lowering in DirectX/dot4add_i8packed.ll
    
 Resolves #99220

inbelic committed:
[HLSL][SPIRV][DXIL] Implement `dot4add_i8packed` intrinsic

    - create a clang built-in in Builtins.td
    - link dot4add_i8packed in hlsl_intrinsics.h
    - add lowering to spirv backend through expansion of operation as
      OPSDot is missing up to SPIRV 1.6 in SPIRVInstructionSelector.cpp
    - add dot4add_i8packed intrinsic to IntrinsicsDirectX.td and mapping
      to DXIL.td op Dot4AddI8Packed

    - add tests for HLSL intrinsic lowering to dx/spv intrinsic in
      dot4add_i8packed.hlsl
    - add tests for sema checks in dot4add_i8packed-errors.hlsl
    - add test of spir-v lowering in SPIRV/dot4add_i8packed.ll
    - add test to dxil lowering in DirectX/dot4add_i8packed.ll

inbelic commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -894,6 +894,16 @@ uint64_t dot(uint64_t3, uint64_t3);
+
+/// \fn int dot4add_i8packed(uint A, uint B, int C)
+
+_HLSL_16BIT_AVAILABILITY(shadermodel, 6.4)
```
```suggestion
_HLSL_AVAILABILITY(shadermodel, 6.4)
```
Whoops, will update with other review comments.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1694,6 +1698,84 @@ bool SPIRVInstructionSelector::selectIntegerDot(Register ResVReg,
   return Result;
 }
 
+// Since pre-1.6 SPIRV has no DotProductInput4x8BitPacked implementation,
+// extract the elements of the packed inputs, multiply them and add the result
+// to the accumulator.
```
This is less efficient code. We should look to see if the instruction is available by checking if the you are targeting spirv1.6 or higher, or the [SPV_KHR_integer_dot_product](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_integer_dot_product.html) extension is enabled. You can get this information from the `SPIRVSubTarget`. I believe it is a member of `SPIRVInstructionSelector` and it is called `STI`.

`STI->canUseExtension(...)`
`STI->isAtLeastSPIRVVer(...)`

inbelic replied:
Okay sounds good thanks. In terms of when the operation is not available, do you agree with the current expansion implementation?

s-perron replied:
I noticed one small problem. I have another comment for that. Otherwise looks good.

s-perron commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/dot4add_i8packed.ll:
```diff
@@ -0,0 +1,48 @@
+; RUN: llc -O0 -mtriple=spirv32v1.3-vulkan-unknown %s -o - | FileCheck %s
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv32v1.3-vulkan-unknown %s -o - -filetype=obj | spirv-val %}
```
You should be using `-mtriple=spirv-vulkan-unknown`. Also add tests that use Vulkan 1.3. `-mtriple=spirv-vulkan1.3-unknown` I believe. Also a test that enables the extension: `--spirv-ext=+SPV_KHR_integer_dot_product`.

You might need changes to SPIRVModuleAnalysis.cpp to get this to work.

inbelic replied:
Updated the testcases to test the different cases. I seem required to use `spirv[32|64]` over `spirv` to meet the `Int8` capability requirements.

s-perron replied:
You should be using `spirv`. You probably need to update `initAvailableCapabilitiesForVulkan` to allow the `Int8` capability. That is core in every version of spir-v, so you can make it always available.

s-perron reviewed (commented):
Thanks for checking with me. To generate better SPIR-V, we should try to use the `OpSDot` instruction when possible. Please use the capability when tageting SPIR-V 1.6 or the extension is available. Thanks.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1694,6 +1698,84 @@ bool SPIRVInstructionSelector::selectIntegerDot(Register ResVReg,
+                  .addUse(GR.getSPIRVTypeID(ResType))
+                  .addUse(I.getOperand(3).getReg())
+                  .addUse(GR.getOrCreateConstInt(i * 8, I, EltType, TII))
+                  .addImm(8)
```
I think you need to use `GR.getOrCreateConstInt(..)`. The [OpBitFieldSExtract](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpBitFieldSExtract) opcodes takes the `count` as an id, not an immediate. This means the spir-v should have

```
%uint_8 = OpConstant %int 8
%f = OpBitFieldSExtract %int_32 %base %uint_0 %uint_8
```

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1694,6 +1698,84 @@ bool SPIRVInstructionSelector::selectIntegerDot(Register ResVReg,
+                  .addUse(GR.getSPIRVTypeID(ResType))
+                  .addUse(Mul)
+                  .addUse(GR.getOrCreateConstInt(0, I, EltType, TII))
+                  .addImm(8)
```
Same here.

inbelic committed:
add SPV_KHR_integer_dot_product flag

we were unable to add this capability to SPIRV target information as a
command line option of llc

inbelic committed:
review comments:

- use OpSDot for the lowering when the target has capabilities to do
so

- fix HLSL_AVAILABILITY

inbelic committed:
review comments

- dont use immediate constant

s-perron commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/dot4add_i8packed.ll:
```diff
@@ -1,48 +1,59 @@
 
+; Test that we use the dot product op when capabilities allow
+
+; CHECK-DOT: %[[#DOT:]] = OpDot %[[#int_32]] %[[#A]] %[[#B]]
```
You should also check that the capability was added to the module. I don't think it is right now.

The first checks should be:

; CHECK-DOT: OpCapability DotProductInput4x8BitPackedKHR
; CHECK-EXT: OpExtension "SPV_KHR_integer_dot_product"

s-perron replied:
The extension should show up only if `-spirv-ext` is used.

inbelic replied:
Right, I believe I have addressed this now. However, I am not 100% sure about correctness. Namely: the use of `addAvailableCaps(...)` in `initAvailableCapabilitiesForOpenCL`. Do I also need to add something similar for the `initAvailableCapabilitiesForVulkan`? Even when I specify using `vulkan` as the 'operating system' in the target triple, it doesn't seem to go down that code path.

s-perron replied:
If the test is calling `initAvailableCapabilitiesForOpenCL` and not `initAvailableCapabilitiesForVulkan`, there is something wrong with the triple. For this change you should not have to change the OpenCL code. You need to use `spirv` and not `spirv32` or `spirv64`.

s-perron replied:
You might also need to add the capabilities that are enabled by the extensions in the Vulkan code as is done for OpenCL. It might be good to more it to `initAvailableCapabilities` since it will be the same for both.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1694,6 +1701,110 @@ bool SPIRVInstructionSelector::selectIntegerDot(Register ResVReg,
+  MachineBasicBlock &BB = *I.getParent();
+
+  Register Dot = MRI->createVirtualRegister(&SPIRV::IDRegClass);
+  bool Result = BuildMI(BB, I, I.getDebugLoc(), TII.get(SPIRV::OpDot))
```
You want to use the [OpSDot](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSDot) and [OpUDot](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpUDot) instructions. If I said OpDot before that was a mistake on my part.

inbelic replied:
Updated. I had to define the ops in `SPIRVInstrInfo`, I noticed there seems to be a mirrored definition in the mlir  codebase (`mlir/Dialect/SPIRV/IR/SPIRVBase.td`). Is it required that I update that now as well?

s-perron replied:
No you do not have to update the mlir code.

inbelic committed:
define and use Op[S|U]Dot

inbelic committed:
add integer_dot_product capabilities

- define the 4 capabilities and add them to OpenCL init when SPIRV
version is 1.6 or greater
  - require these capabilities during analysis of OpSDot or OpUDot
instructions
  - verify in test case that the capability/extensions are correctly
emitted

s-perron commented on llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp:
```diff
@@ -1218,6 +1223,17 @@ void addInstrRequirements(const MachineInstr &MI,
+    if (ST.canUseExtension(SPIRV::Extension::SPV_KHR_integer_dot_product)) {
+      Reqs.addExtension(SPIRV::Extension::SPV_KHR_integer_dot_product);
+    }
+    Reqs.addCapability(SPIRV::Capability::DotProductKHR);
+    Reqs.addCapability(SPIRV::Capability::DotProductInputAllKHR);
+    Reqs.addCapability(SPIRV::Capability::DotProductInput4x8BitKHR);
+    Reqs.addCapability(SPIRV::Capability::DotProductInput4x8BitPackedKHR);
```
You don't need all of the requirements. This can be determined based on the type of the operands:

> Vector 1 and Vector 2 must be either 32-bit integers (enabled by DotProductInput4x8BitPackedKHR) or vectors of integer type (enabled by DotProductInput4x8BitKHR or DotProductInputAllKHR).

inbelic committed:
move out common capabilities

inbelic committed:
add extension and capabilities to vulkan env

inbelic committed:
only add required capabilities

pow2clk commented on clang/test/CodeGenHLSL/builtins/dot4add_i8packed.hlsl:
```diff
@@ -0,0 +1,17 @@
+// RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -triple \
```
202x is the default. Is there a reason this requires 2021?

pow2clk commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2527,6 +2640,11 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_sdot:
     return selectIntegerDot(ResVReg, ResType, I);
+  case Intrinsic::spv_dot4add_i8packed:
+    if (STI.canUseExtension(SPIRV::Extension::SPV_KHR_integer_dot_product) ||
```
Nice to have this option. Now that we do, it would be nice to update the regular dot product to use this in the integer cases, but that can be left for a different change.

inbelic replied:
Created for tracking here: https://github.com/llvm/llvm-project/issues/114632

pow2clk reviewed (approved):
Looks good! I learned a couple things about SPIR-V by verifying this. 

Just one quibble about the language version and a suggested followup.

inbelic committed:
remove unneeded flags

s-perron reviewed (approved):
LGTM. I just want someone on the OpenCL side to make sure they are okay with the changes for OpenCL.

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp:
```diff
@@ -1218,6 +1224,32 @@ void addInstrRequirements(const MachineInstr &MI,
     break;
+  case SPIRV::OpSDot:
+  case SPIRV::OpUDot: {
+    if (ST.canUseExtension(SPIRV::Extension::SPV_KHR_integer_dot_product))
```
Let's please move from the switch's case code to a separate function to make the switch code a bit cleaner.

VyacheslavLevytskyy commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/dot4add_i8packed.ll:
```diff
@@ -0,0 +1,65 @@
+; RUN: llc -O0 -mtriple=spirv-unknown-unknown %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-EXP
```
The default version will change and this test will fail. It's better to insert a version number explicitly here to be sure that it's not 1.6

VyacheslavLevytskyy commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/dot4add_i8packed.ll:
```diff
@@ -0,0 +1,65 @@
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv-unknown-unknown %s -o - -filetype=obj | spirv-val %}
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv1.6-unknown-unknown %s -o - -filetype=obj | spirv-val %}
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv-unknown-unknown -spirv-ext=+SPV_KHR_integer_dot_product %s -o - -filetype=obj | spirv-val %}
+
```
Please add `-verify-machineinstrs` command line argument to all `llc`'s.

inbelic committed:
move dot product requirements out of switch

inbelic committed:
review comments

  - use getRegClass instead of IDRegClass
  - remove unneeded check of isvulkan/opencl
  - fix testcase to use fixed version and -verify-machineinstrs
  - fix Result bool to be from and to or

inbelic committed:
change name to use not KHR version

inbelic commented:
Regarding the comments about differentiating between `OpSDot/DotProduct` and `OpSDotKHR/DotProductKHR`:

I agree that these are different ops/capabilities in the SPIR-V spec.

However from a quick attempt to implement this, I found that since the Capabilities share the same bit value of (`6016, 6017,...`), we do not have a way to differentiate between these capabilities when checking requirements and adding available capabilities. This is because `SPIRV::Capability::DotProduct == SPIRV::Capability::DotProductKHR = 6019`.

I think implementing a way to differentiate capabilities that have the same value is out of scope for this pr. So I would propose that we file a follow-up issue to deal with that and either:
- Merge as-is
- Remove the extension based changes from this pr, so only use `OpSDot` and the `DotProduct` capability when we have a version that fits
- Remove the version based changes from this pr, so only use `OpSDotKHR` and the `DotProductKHR` capability when the extension is provided
- Remove both optimized versions and only lower to the expansion

WDYT?

inbelic committed:
self-review:

    - add minimum version to capabilities

s-perron commented:
> Regarding the comments about differentiating between `OpSDot/DotProduct` and `OpSDotKHR/DotProductKHR`:
> 
> I agree that these are different ops/capabilities in the SPIR-V spec.
> 
> However from a quick attempt to implement this, I found that since the Capabilities share the same bit value of (`6016, 6017,...`), we do not have a way to differentiate between these capabilities when checking requirements and adding available capabilities. This is because `SPIRV::Capability::DotProduct == SPIRV::Capability::DotProductKHR = 6019`.
> 
> I think implementing a way to differentiate capabilities that have the same value is out of scope for this pr. So I would propose that we file a follow-up issue to deal with that and either:
> 
> * Merge as-is
> * Remove the extension based changes from this pr, so only use `OpSDot` and the `DotProduct` capability when we have a version that fits
> * Remove the version based changes from this pr, so only use `OpSDotKHR` and the `DotProductKHR` capability when the extension is provided
> * Remove both optimized versions and only lower to the expansion
> 
> WDYT?

I'll check with the authors of spirv-as and spirv-dis, but here are my thoughts.

This distinction makes no difference in the binary format, which is what is defined by the SPIR-V specification. The SPIR-V specification does not define the human readable format. See https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_example.

The textual format ouput by glslang by default is actually very different. It is just that the format used by spirv tools happens to have one. If we use Spirv-Tools as the guide, then we can use either one interchangeably. Here are two examples:

1. Spirv-dis will always output the non-KHR string. `OpSDotKHR` is used initially, but when disassembled it uses `OpSDot`: https://godbolt.org/z/bPf67vnss
2. Spirv-as accepts the non-KHR string: https://godbolt.org/z/Mf8b9x8dc

I think it is fair for the llvm backend to do the same. The only time people will have a problem is when they use llvm to generate the asm, and try to generate a binary using an older spirv-as. That is easily fixed by generating the binary in llvm.

dneto0 commented:
Hi, former SPIR WG chair here, and long time maintainer of spirv-as and spirv-dis.

@s-perron is correct: only the numbers in the binary are normative.

> I think implementing a way to differentiate capabilities that have the same value is out of scope for this pr.

I recommend not addressing this, ever.   It's not worth it. 

spirv-as will map both the old and new string names to the number; it is driven from the grammar file and knows all registered names for numbers.  (e.g. see [here](https://github.com/KhronosGroup/SPIRV-Headers/blob/cb6b2c32dbfc3257c1e9142a116fe9ee3d9b80a2/include/spirv/unified1/spirv.core.grammar.json#L16408) for the declaration that capability strings `DotProduct` and `DotProductKHR` both map to 6019.

spirv-dis maps the number to what it considers the best name, which is the string listed for the `enumerant`, field and not any of the `alias` names.  By convention, the most standardized name is put in that slot. So the standard core name is better than the KHR, which is better than an EXT, which is better than a vendor-specific name.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1694,6 +1701,112 @@ bool SPIRVInstructionSelector::selectIntegerDot(Register ResVReg,
+  MachineBasicBlock &BB = *I.getParent();
+
+  auto DotOp = Signed ? SPIRV::OpSDot : SPIRV::OpUDot;
+  Register Dot = MRI->createVirtualRegister(&SPIRV::IDRegClass);
```
That is good to know. I had not learned about the different register classes yet. Thanks for catching that.

inbelic committed:
fix clang formatting

inbelic committed:
Merge branch 'main' into inbelic/packed-dot-i8

farzonl commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -934,6 +934,16 @@ uint64_t dot(uint64_t3, uint64_t3);
+
+_HLSL_AVAILABILITY(shadermodel, 6.4)
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_dot4add_i8packed)
+int dot4add_i8packed(unsigned int, unsigned int, int);
```
is there a reason you didn't do `uint` here? Its the same, so just curious.

inbelic replied:
Whoops no. A discrepancy on my part.
